File: backend/skills/readiness.go
package skills

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/enough/enough/backend/enoughhome"
)

type SkillReadinessStatus string

const (
	ReadinessAvailable   SkillReadinessStatus = "available"
	ReadinessSetupNeeded SkillReadinessStatus = "setup_needed"
	ReadinessUnsupported SkillReadinessStatus = "unsupported"
)

type RequiredEnvVar struct {
	Name        string `json:"name"`
	Prompt      string `json:"prompt"`
	Help        string `json:"help,omitempty"`
	RequiredFor string `json:"required_for,omitempty"`
	Optional    bool   `json:"optional,omitempty"`
}

type CollectSecretEntry struct {
	EnvVar      string `json:"env_var"`
	Prompt      string `json:"prompt"`
	ProviderURL string `json:"provider_url,omitempty"`
	Secret      bool   `json:"secret"`
}

type SetupBlock struct {
	Help           *string               `json:"help"`
	CollectSecrets []CollectSecretEntry `json:"collect_secrets"`
}

var envVarNameRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

func LoadEnoughEnv() map[string]string {
	envMap := make(map[string]string)
	home := enoughhome.HomeDir()
	envPath := filepath.Join(home, ".env")
	data, err := os.ReadFile(envPath)
	if err != nil {
		return envMap
	}
	lines := strings.Split(string(data), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		if len(parts) == 2 {
			k := strings.TrimSpace(parts[0])
			v := strings.TrimSpace(parts[1])
			v = strings.Trim(v, `"'`)
			envMap[k] = v
		}
	}
	return envMap
}

func isEnvVarSet(name string, envMap map[string]string) bool {
	if os.Getenv(name) != "" {
		return true
	}
	if v, ok := envMap[name]; ok && v != "" {
		return true
	}
	return false
}

func normalizeSetupMetadata(fm map[string]interface{}) SetupBlock {
	setupVal, ok := fm["setup"]
	if !ok || setupVal == nil {
		return SetupBlock{CollectSecrets: []CollectSecretEntry{}}
	}
	setupMap, ok := setupVal.(map[string]interface{})
	if !ok {
		return SetupBlock{CollectSecrets: []CollectSecretEntry{}}
	}

	var helpText *string
	if h, ok := setupMap["help"].(string); ok && strings.TrimSpace(h) != "" {
		trimmed := strings.TrimSpace(h)
		helpText = &trimmed
	}

	var collectSecrets []CollectSecretEntry
	csVal := setupMap["collect_secrets"]
	if csVal != nil {
		var rawList []interface{}
		if singleMap, ok := csVal.(map[string]interface{}); ok {
			rawList = []interface{}{singleMap}
		} else if list, ok := csVal.([]interface{}); ok {
			rawList = list
		}
		for _, item := range rawList {
			itemMap, ok := item.(map[string]interface{})
			if !ok {
				continue
			}
			var envVar string
			if ev, ok := itemMap["env_var"].(string); ok {
				envVar = strings.TrimSpace(ev)
			}
			if envVar == "" {
				continue
			}
			prompt := fmt.Sprintf("Enter value for %s", envVar)
			if p, ok := itemMap["prompt"].(string); ok && strings.TrimSpace(p) != "" {
				prompt = strings.TrimSpace(p)
			}
			var providerURL string
			if pu, ok := itemMap["provider_url"].(string); ok {
				providerURL = strings.TrimSpace(pu)
			} else if u, ok := itemMap["url"].(string); ok {
				providerURL = strings.TrimSpace(u)
			}

			secret := true
			if sec, ok := itemMap["secret"]; ok {
				if secBool, ok := sec.(bool); ok {
					secret = secBool
				}
			}
			collectSecrets = append(collectSecrets, CollectSecretEntry{
				EnvVar:      envVar,
				Prompt:      prompt,
				ProviderURL: providerURL,
				Secret:      secret,
			})
		}
	}

	return SetupBlock{
		Help:           helpText,
		CollectSecrets: collectSecrets,
	}
}

func getRequiredEnvironmentVariables(fm map[string]interface{}) []RequiredEnvVar {
	setup := normalizeSetupMetadata(fm)
	var required []RequiredEnvVar
	seen := make(map[string]bool)

	appendRequired := func(entry map[string]interface{}) {
		var name string
		if n, ok := entry["name"].(string); ok {
			name = strings.TrimSpace(n)
		} else if ev, ok := entry["env_var"].(string); ok {
			name = strings.TrimSpace(ev)
		}
		if name == "" || seen[name] {
			return
		}
		if !envVarNameRe.MatchString(name) {
			return
		}

		prompt := fmt.Sprintf("Enter value for %s", name)
		if p, ok := entry["prompt"].(string); ok && strings.TrimSpace(p) != "" {
			prompt = strings.TrimSpace(p)
		}

		var helpText string
		if h, ok := entry["help"].(string); ok && strings.TrimSpace(h) != "" {
			helpText = strings.TrimSpace(h)
		} else if pu, ok := entry["provider_url"].(string); ok && strings.TrimSpace(pu) != "" {
			helpText = strings.TrimSpace(pu)
		} else if u, ok := entry["url"].(string); ok && strings.TrimSpace(u) != "" {
			helpText = strings.TrimSpace(u)
		} else if setup.Help != nil {
			helpText = *setup.Help
		}

		var requiredFor string
		if rf, ok := entry["required_for"].(string); ok {
			requiredFor = strings.TrimSpace(rf)
		}

		var optional bool
		if opt, ok := entry["optional"].(bool); ok {
			optional = opt
		}

		seen[name] = true
		required = append(required, RequiredEnvVar{
			Name:        name,
			Prompt:      prompt,
			Help:        helpText,
			RequiredFor: requiredFor,
			Optional:    optional,
		})
	}

	// 1. required_environment_variables
	reqRaw := fm["required_environment_variables"]
	if reqRaw != nil {
		var rawList []interface{}
		if singleStr, ok := reqRaw.(string); ok {
			rawList = []interface{}{singleStr}
		} else if singleMap, ok := reqRaw.(map[string]interface{}); ok {
			rawList = []interface{}{singleMap}
		} else if list, ok := reqRaw.([]interface{}); ok {
			rawList = list
		}
		for _, item := range rawList {
			if s, ok := item.(string); ok {
				appendRequired(map[string]interface{}{"name": s})
			} else if m, ok := item.(map[string]interface{}); ok {
				appendRequired(m)
			}
		}
	}

	// 2. setup.collect_secrets
	for _, item := range setup.CollectSecrets {
		appendRequired(map[string]interface{}{
			"name":         item.EnvVar,
			"prompt":       item.Prompt,
			"provider_url": item.ProviderURL,
		})
	}

	// 3. legacy prerequisites.env_vars
	if prereqs, ok := fm["prerequisites"].(map[string]interface{}); ok {
		if envVarsVal := prereqs["env_vars"]; envVarsVal != nil {
			var list []string
			if s, ok := envVarsVal.(string); ok && strings.TrimSpace(s) != "" {
				list = []string{strings.TrimSpace(s)}
			} else if l, ok := envVarsVal.([]interface{}); ok {
				for _, it := range l {
					if s, ok := it.(string); ok && strings.TrimSpace(s) != "" {
						list = append(list, strings.TrimSpace(s))
					}
				}
			}
			for _, ev := range list {
				appendRequired(map[string]interface{}{"name": ev})
			}
		}
	}

	return required
}

func CheckSkillsRequirements() bool {
	return true
}

File: backend/skills/readiness.ts
// PORT: backend/skills/readiness.go

import fs from "node:fs";
import path from "node:path";
import { HomeDir } from "./paths";
import { toStringList } from "./frontmatter";

export type SkillReadinessStatus = "available" | "setup_needed" | "unsupported";

export const ReadinessAvailable: SkillReadinessStatus = "available";
export const ReadinessSetupNeeded: SkillReadinessStatus = "setup_needed";
export const ReadinessUnsupported: SkillReadinessStatus = "unsupported";

export interface RequiredEnvVar {
  Name: string;
  Prompt: string;
  Help?: string;
  RequiredFor?: string;
  Optional?: boolean;
}

export interface CollectSecretEntry {
  EnvVar: string;
  Prompt: string;
  ProviderURL?: string;
  Secret: boolean;
}

export interface SetupBlock {
  Help: string | null;
  CollectSecrets: CollectSecretEntry[];
}

const envVarNameRe = /^[A-Za-z_][A-Za-z0-9_]*$/;

export function LoadHollowEnv(): Record<string, string> {
  const envMap: Record<string, string> = {};
  const home = HomeDir();
  const envPath = path.join(home, ".env");
  let data: string;
  try {
    data = fs.readFileSync(envPath, "utf8");
  } catch {
    return envMap;
  }
  const lines = data.split("\n");
  for (let line of lines) {
    line = line.trim();
    if (line === "" || line.startsWith("#")) {
      continue;
    }
    const parts = line.split("=");
    if (parts.length >= 2) {
      const k = parts[0].trim();
      let v = parts.slice(1).join("=").trim();
      v = v.replace(/^["']|["']$/g, ""); // Strip leading/trailing quotes
      envMap[k] = v;
    }
  }
  return envMap;
}

export function isEnvVarSet(name: string, envMap: Record<string, string>): boolean {
  if (process.env[name] !== undefined && process.env[name] !== "") {
    return true;
  }
  if (envMap[name] !== undefined && envMap[name] !== "") {
    return true;
  }
  return false;
}

export function normalizeSetupMetadata(fm: Record<string, any>): SetupBlock {
  const setupVal = fm["setup"];
  if (setupVal === undefined || setupVal === null) {
    return { Help: null, CollectSecrets: [] };
  }
  const setupMap = setupVal as Record<string, any>;
  if (typeof setupMap !== "object") {
    return { Help: null, CollectSecrets: [] };
  }

  let helpText: string | null = null;
  if (typeof setupMap["help"] === "string" && setupMap["help"].trim() !== "") {
    helpText = setupMap["help"].trim();
  }

  const collectSecrets: CollectSecretEntry[] = [];
  const csVal = setupMap["collect_secrets"];
  if (csVal !== undefined && csVal !== null) {
    let rawList: any[] = [];
    if (Array.isArray(csVal)) {
      rawList = csVal;
    } else if (typeof csVal === "object") {
      rawList = [csVal];
    }

    for (const item of rawList) {
      if (!item || typeof item !== "object") {
        continue;
      }
      let envVar = "";
      if (typeof item["env_var"] === "string") {
        envVar = item["env_var"].trim();
      }
      if (envVar === "") {
        continue;
      }
      let prompt = `Enter value for ${envVar}`;
      if (typeof item["prompt"] === "string" && item["prompt"].trim() !== "") {
        prompt = item["prompt"].trim();
      }
      let providerURL = "";
      if (typeof item["provider_url"] === "string") {
        providerURL = item["provider_url"].trim();
      } else if (typeof item["url"] === "string") {
        providerURL = item["url"].trim();
      }

      let secret = true;
      if (item["secret"] !== undefined) {
        if (typeof item["secret"] === "boolean") {
          secret = item["secret"];
        }
      }
      collectSecrets.push({
        EnvVar: envVar,
        Prompt: prompt,
        ProviderURL: providerURL || undefined,
        Secret: secret,
      });
    }
  }

  return {
    Help: helpText,
    CollectSecrets: collectSecrets,
  };
}

export function getRequiredEnvironmentVariables(fm: Record<string, any>): RequiredEnvVar[] {
  const setup = normalizeSetupMetadata(fm);
  const required: RequiredEnvVar[] = [];
  const seen = new Set<string>();

  const appendRequired = (entry: Record<string, any>) => {
    let name = "";
    if (typeof entry["name"] === "string") {
      name = entry["name"].trim();
    } else if (typeof entry["env_var"] === "string") {
      name = entry["env_var"].trim();
    }
    if (name === "" || seen.has(name)) {
      return;
    }
    if (!envVarNameRe.test(name)) {
      return;
    }

    let prompt = `Enter value for ${name}`;
    if (typeof entry["prompt"] === "string" && entry["prompt"].trim() !== "") {
      prompt = entry["prompt"].trim();
    }

    let helpText = "";
    if (typeof entry["help"] === "string" && entry["help"].trim() !== "") {
      helpText = entry["help"].trim();
    } else if (typeof entry["provider_url"] === "string" && entry["provider_url"].trim() !== "") {
      helpText = entry["provider_url"].trim();
    } else if (typeof entry["url"] === "string" && entry["url"].trim() !== "") {
      helpText = entry["url"].trim();
    } else if (setup.Help !== null) {
      helpText = setup.Help;
    }

    let requiredFor = "";
    if (typeof entry["required_for"] === "string") {
      requiredFor = entry["required_for"].trim();
    }

    let optional = false;
    if (typeof entry["optional"] === "boolean") {
      optional = entry["optional"];
    }

    seen.add(name);
    required.push({
      Name: name,
      Prompt: prompt,
      Help: helpText || undefined,
      RequiredFor: requiredFor || undefined,
      Optional: optional,
    });
  };

  // 1. required_environment_variables
  const reqRaw = fm["required_environment_variables"];
  if (reqRaw !== undefined && reqRaw !== null) {
    let rawList: any[] = [];
    if (typeof reqRaw === "string") {
      rawList = [reqRaw];
    } else if (Array.isArray(reqRaw)) {
      rawList = reqRaw;
    } else if (typeof reqRaw === "object") {
      rawList = [reqRaw];
    }
    for (const item of rawList) {
      if (typeof item === "string") {
        appendRequired({ name: item });
      } else if (item && typeof item === "object") {
        appendRequired(item);
      }
    }
  }

  // 2. setup.collect_secrets
  for (const item of setup.CollectSecrets) {
    appendRequired({
      name: item.EnvVar,
      prompt: item.Prompt,
      provider_url: item.ProviderURL,
    });
  }

  // 3. legacy prerequisites.env_vars
  const prereqs = fm["prerequisites"];
  if (prereqs && typeof prereqs === "object") {
    const envVarsVal = prereqs["env_vars"];
    if (envVarsVal !== undefined && envVarsVal !== null) {
      let list: string[] = [];
      if (typeof envVarsVal === "string" && envVarsVal.trim() !== "") {
        list = [envVarsVal.trim()];
      } else if (Array.isArray(envVarsVal)) {
        list = toStringList(envVarsVal);
      }
      for (const ev of list) {
        appendRequired({ name: ev });
      }
    }
  }

  return required;
}

export function CheckSkillsRequirements(): boolean {
  return true;
}

/*
PORT STATUS
source path: backend/skills/readiness.go
source lines: 256
draft lines: 255
confidence: high
status: phase_b_compile
*/
